### ISSUE TITLE
cachedImplicit "in the same compilation unit"

### ISSUE THREAD

**Issue by fommil** [milessabin/shapeless#380]
when I try to use `cachedImplicit` I am getting this error

```
FamilyFormatsSpec.scala:70: macro implementation not found: cachedImplicit
[error] (the most common reason for that is that you cannot use macro implementations in the same compilation run that defines them)
[error]   implicit val SimpleTraitFormat: RootJsonFormat[SimpleTrait] = cachedImplicit
```

e.g. in this file

https://github.com/fommil/ensime-server/blob/jerk/spray-json-shapeless/src/test/scala/org/ensime/json/FamilyFormatsSpec.scala#L69

I've tried it in a few different places and get the same error each time. Is there some rule I must obey?

**Comment by milessabin** [comment]
By a spooky coincidence I'm also experience this just now ... currently investigating ...

**Comment by fommil** [comment]
another thing confuses me about this: if I create a `cachedImplicit` for a `LabelledGeneric[T]` will that be used by methods that need a `LabelledGeneric.Aux[T, H]`?

**Comment by milessabin** [comment]
You'd need to cache an instance with the the full type of the `Generic`, including its refinement (ie. the `type Repr = ...` which is captured by `Generic.Aux[T, R]` but discarded by `Generic[T]`.

That makes `cachedImplicit` quite hard to use here. I think instead you'll have to do something like this,
```scala
val gen = {
  def cachedGen = ??? // shadow following implicit defn
  Generic[T]
}
implicit val cachedGen: Generic[T, gen.Repr] = gen // type annotation on implicit cachedGen
```
Unfortunately I don't think there's anything I can do here to improve on this. The problem is that we need to typecheck `Generic[T]` before we can use `gen.Repr` on the LHS of the definition of `cachedGen`, so a macro on the RHS of that definition isn't going to help us.

Could you give the above a try ... even though it's a bit awkward I think it will have a positive effect on compile times and footprint, so I think it is worth exploring.

**Comment by milessabin** [comment]
Fixed in 619664b1cd6d0ced18aba2a55985c7318968a993.

**Comment by fommil** [comment]
in the above example, where do I used `cachedImplicit`?

**Comment by fommil** [comment]
btw, this is a fuller example of how I'm using this and the thing that I believe is blowing up the presentation compiler https://github.com/ensime/ensime-server/issues/954

```scala
sealed trait StringEnum {
    def label: String
  }
case class Flooma(label: String) extends StringEnum
case class Blam(label: String) extends StringEnum

// http://stackoverflow.com/questions/29676204
object SealedTraitImpls {
  class NameHelper[A] {
    def apply[C <: Coproduct, K <: HList]()(
      implicit
      gen: LabelledGeneric.Aux[A, C],
      keys: ops.union.Keys.Aux[C, K],
      toSet: ops.hlist.ToTraversable.Aux[K, Set, Symbol]
    ): Set[String] = toSet(keys()).map(_.name)
  }
  def names[A] = new NameHelper[A]
}

object StringEnumMapping {
  // https://github.com/ensime/ensime-server/issues/954
  val gen = {
    def cachedGen = ???
    LabelledGeneric[StringEnum]
  }
  implicit val cachedGen: LabelledGeneric.Aux[StringEnum, gen.Repr] = gen

  // could probably do with implicit caches for Keys and the Traversable as well
  val values = SealedTraitImpls.names[StringEnum]().toList.sorted
}
```

**Comment by fommil** [comment]
btw, the SNAPSHOT `cachedImplicit` works fine in the spray-json-shapeless code that was failing yesterday (although I'm still not sure if I need it there).

**Comment by milessabin** [comment]
@fommil the example was an alternative to using `cachedImplicit` in cases where you need to cache an implicit with a refinement (ie. `{ type Repr = ... }` which it would be object-defeating to have to write explicitly as a type annotation on the LHS.

**Comment by fommil** [comment]
ok. I'm currently refactoring this particular example so that I can cache an implicit `NameHelper` instead.

**Comment by milessabin** [comment]
Does that `StringEnum` that you just posted kill the presentation compiler exactly as written?

**Comment by fommil** [comment]
I've not got a minimal failing case yet, but the problems started to happen around the time when I started using this `NameHelper` to list the implementation names. Seems fine to use with the normal compiler.

**Comment by milessabin** [comment]
So I'm not clear ... is that code problematic or not?

**Comment by fommil** [comment]
yes, problematic code (I'm 99% sure... it's a bloody nightmare formalising an infinite loop in the presentation compiler).

**Comment by milessabin** [comment]
That exact example?

**Comment by fommil** [comment]
gimme 10 minutes and I'll confirm. It's not **exactly** the above (the actual `StringEnum` has about 15 impls in my real code)

**Comment by fommil** [comment]
I can't reproduce the problem just yet. Getting the presentation compiler into a bad state is apparently very difficult when I want to do it, yet inevitable when I just want it to tell me the damn type at point. :smile:

I'll open a new ticket if I can reproduce reliably (which will most likely be as an ENSIME integration test).

**Comment by milessabin** [comment]
Awesome ... thanks :-)
